sql/inspect: add DETACHED option to run INSPECT without waiting

spilchen opened the pull request:
Adds support for running `INSPECT ... WITH OPTIONS (DETACHED)`, which submits the inspection as a background job and returns immediately with the job ID in a notice. This allows INSPECT to run inside a transaction.

**Note to reviewers:**
This was added to support running INSPECT within the random schema workload, which executes statements inside transactions. Previously, INSPECT could not run inside a transaction. It also gives us async behaviour. While expanding INSPECT test coverage, I have wanted to kick off an async run of INSPECT. Until now, I worked around it by using `statement_timeout` to force the job to run in the background so we could later wait on it. The detached option provides a cleaner approach. It is similar to how commands like BACKUP support an async mode.
 
Informs: #155483
Epic: CRDB-55075

Release note (sql change): INSPECT supports a DETACHED option to run the operation without waiting for it.

bghal reviewed (approved):
> @bghal reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155774)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
nice! this will be useful for end-users as well

one thing to double-check -- does the NOTICE added here also get displayed still? https://github.com/cockroachdb/cockroach/blob/88e72fea0d92563aa3e26725ee62e51336049c23/pkg/sql/conn_executor.go#L4241

the reason i ask is because i added logic to the logictest framework to ignore this notice since it makes the `query T noticetrace` test cases non-deterministic: https://github.com/cockroachdb/cockroach/blob/11f6cddf0442be37c10fc5418c27e804a8ce99bf/pkg/sql/logictest/logic.go#L1308

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155774#-:-OcBJi3tBb0ZjWuFgYKB:b4wlcml)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)

___
*[`pkg/sql/inspect/inspect_plan.go` line 248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155774#-OcBKoQf5JhuPOcVj-PU:-OcBKoQf5JhuPOcVj-PV:b-1lqewd) ([raw file](https://github.com/cockroachdb/cockroach/blob/61fa8df1a99d2490cad5bde3c040b4fb7a6066c5/pkg/sql/inspect/inspect_plan.go#L248)):*
> ```Go
> 		},
> 		Progress:      jobspb.InspectProgress{},
> 		CreatedBy:     nil,
> ```

nit: why is CreatedBy nil?


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
That notice doesn't show up for INSPECT. We don't use the same schema changer wait function in the conn executor (`waitForTxnJobs`). INSPECT uses `AwaitCompletion` to do the wait.

I do like the UX change in the notice you added. Otherwise you have to spend time to hunt for the job ID. I may include a similar notice in this change for that case.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155774#-:-OcBkWM-2dPOjPfLu70C:bjeawcd)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/inspect/inspect_plan.go` line 248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/155774#-OcBKoQf5JhuPOcVj-PU:-OcBoH5_5U0Gb4GqEa6q:b-au5jg3) ([raw file](https://github.com/cockroachdb/cockroach/blob/61fa8df1a99d2490cad5bde3c040b4fb7a6066c5/pkg/sql/inspect/inspect_plan.go#L248)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: why is CreatedBy nil?
</blockquote></details>

This is a copy+paste from the inspect job creation in `pkg/sql/inspect_plan.go`. Note: we will eventually remove this duplication when `pkg/sql/inspect_plan.go` is moved to the `pkg/sql/inspect` package (needs scrub removal first).

It is my understanding that this is set for when jobs are started by an automated process (e.g. scheduled backup). 


<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
That one was for the DETACHED case only. I'll use `SendClientNotice'.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/155774#-:-OcC6QF43vTt_LWc-n7h:b-ka5lum)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @bghal and @rafiss)



<!-- Sent from Reviewable.io -->

spilchen committed:
sql/inspect: add DETACHED option to run INSPECT without waiting

Adds support for running `INSPECT ... WITH OPTIONS (DETACHED)`, which
submits the inspection as a background job and returns immediately with
the job ID in a notice. This allows INSPECT to run inside a transaction.

This also adds NOTICES for the INSPECT command (both detached and
non-detached) to identify the job number.

Informs: #155483
Epic: CRDB-55075

Release note (sql change): INSPECT supports a DETACHED option to run the
operation without waiting for it.

spilchen commented:
TFTRs!

bors r+
